File: stack/844.backspace-string-compare.py
#
# @lc app=leetcode id=844 lang=python3
#
# [844] Backspace String Compare
#

# @lc code=start
from typing import List


class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        s = self.realyString(S)
        t = self.realyString(T)
        return s == t

    def realyString(self, originString: str) -> List[str]:
        result: List[str] = []
        for i in originString:
            if i == "#":
                if len(result) > 0:
                    result.pop()
            else:
                result.append(i)
        return result


# @lc code=end

File: stack/844.backspace-string-compare.go
package main

import (
	"fmt"
	"reflect"
)

/*
 * @lc app=leetcode id=844 lang=golang
 *
 * [844] Backspace String Compare
 */

// @lc code=start
func backspaceCompare(S string, T string) bool {
	s := realString(S)
	t := realString(T)
	return reflect.DeepEqual(s, t)
}

func realString(s string) []rune {
	result := []rune{}
	for _, i := range s {
		if i == '#' {
			if len(result) > 0 {
				result = result[:len(result)-1]
			}
		} else {
			result = append(result, i)
		}
	}
	return result
}

// @lc code=end

func main() {
	fmt.Println(backspaceCompare("ab#c", "ad#c"))
	fmt.Println(backspaceCompare("ab#c", "ad##c"))
}
